internal/rbd/controllerserver.go

```diff
@@ -1397,6 +1394,39 @@ func (cs *ControllerServer) DeleteSnapshot(
 	return &csi.DeleteSnapshotResponse{}, nil
 }
 
+// cleanUpImageAndSnapReservation cleans up the image from the trash and
+// snapshot reservation in rados OMAP.
+func cleanUpImageAndSnapReservation(ctx context.Context, rbdSnap *rbdSnapshot, cr *util.Credentials) error {
+	rbdVol := generateVolFromSnap(rbdSnap)
+	err := rbdVol.Connect(cr)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer rbdVol.Destroy()
+
+	err = rbdVol.openIoctx()
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	// cleanup the image from trash if the error is image not found.
+	err = rbdVol.ensureImageCleanup(ctx)
```

Comment:
maybe make `ensureImageCleanup()` a function  of `rbdImage` so that you can call it on the `rbdSnap` directly?

Reply:
This needs more changes than expected as i need to update multiple functions which am not touching in this PR. will revisit when working on the codebase cleanup.

Reply:
ok, thanks!